contracts/btc-finality/src/finality.rs

```diff
@@ -24,19 +24,126 @@ use k256::sha2::{Digest, Sha256};
 use std::cmp::max;
 use std::collections::HashSet;
 
+// The maximum number of blocks into the future
+// that a public randomness commitment start height can target. This limit prevents abuse by capping
+// the size of the commitments index, protecting against potential memory exhaustion
+// or performance degradation caused by excessive future commitments.
+const MAX_PUB_RAND_COMMIT_OFFSET: u64 = 160_000;
+
+const COMMITMENT_LENGTH_BYTES: usize = 32;
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum PubRandCommitError {
+    #[error("Empty FP BTC PubKey")]
+    EmptyFpBtcPubKey,
+    #[error("Commitment must be {COMMITMENT_LENGTH_BYTES} bytes, got {0}")]
+    BadCommitmentLength(usize),
+    #[error("public rand commit start block height: {0} is equal or higher than start_height + num_pub_rand ({1})")]
+    OverflowInBlockHeight(u64, u64),
+    #[error("Empty signature")]
+    EmptySignature,
+}
+
+// https://github.com/babylonlabs-io/babylon/blob/49972e2d3e35caf0a685c37e1f745c47b75bfc69/x/finality/types/tx.pb.go#L36
+pub struct PublicRandomnessCommitMsg {
+    pub fp_btc_pk_hex: String,
+    pub start_height: u64,
+    pub num_pub_rand: u64,
+    pub commitment: Vec<u8>,
+    pub sig: Vec<u8>,
+}
+
+impl PublicRandomnessCommitMsg {
+    // https://github.com/babylonlabs-io/babylon/blob/49972e2d3e35caf0a685c37e1f745c47b75bfc69/x/finality/types/msg.go#L161
```

Comment:
maybe better to put the ref on the top of PublicRandomnessCommitMsg as the ref covers the entire impl

Reply:
You meant the entire impl of `PublicRandomnessCommitMsg` structure? That's not true, this permalink is specifically for the function `ValidateBasic`.